### PR TITLE
Fix upsample_bilinear to respect align_corner argument

### PR DESCRIPTION
**Pull request by BowenBao**
Fixes https://github.com/microsoft/onnxscript/issues/1159#issuecomment-1888089502 which indeed turns out to be a problem uncovered by PyTorch CI https://github.com/pytorch/pytorch/actions/runs/7508784822/job/20445196351?pr=117314.

> Fixes `align_corner` default value. The default value from pytorch signature is `False` https://pytorch.org/docs/stable/generated/torch.nn.Upsample.html#torch.nn.Upsample.
> That said, it shouldn't matter since `align_corner` in aten signature in `native_functions.yaml` is a required argument, so in practice this function will never be invoked w/o `align_corner`.  

Above is outdated. The case is more complicated. https://github.com/microsoft/onnxscript/pull/1254#discussion_r1453948494.

In short this PR fixes the torchlib op signature to match with aten spec, and updates input wrangler for test case to bridge from sample test inputs for function `torch.nn.functional.upsample_bilinear`.

### CODE REVIEW

**Comment by BowenBao** [comment]
@justinchuby The reason why it passes on onnxscript side is that we are not testing the `nn.functional.interpolate` test case from `common_methods_invocations.py`, which fails on pytorch side. I don't know how to properly add the test case for this, it is a large test case with various variants. The `upsample` test case we test on probably didn't have `align_corner` flag variation in its sample inputs.

We need more e2e pytorch export tests when adding ops in torchlib to ensure it works.

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
Thanks!

> We need more e2e pytorch export tests when adding ops in torchlib to ensure it works.

Do you have suggestions? Currently we rely on tests in PyTorch for integration tests.

File: onnxscript/function_libs/torch_lib/ops/nn.py [hunk 1]
```diff
@@ -2298,16 +2298,21 @@ def aten_upsample_bilinear2d(
...
 
+    coordinate_transformation_mode = "align_corners" if align_corners else "pytorch_half_pixel"
     if output_size is not None:
-        result = _aten_upsample_bilinear2d_output_size(self, output_size)
+        result = _aten_upsample_bilinear2d_output_size(
+            self, output_size, coordinate_transformation_mode
+        )
     else:
         assert scales_h is not None
         assert scales_h == scales_w, f"scale_h({scales_h}) != scale_w({scales_w})"
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/nn.py:2312]: Is this still valid?

File: onnxscript/function_libs/torch_lib/ops/nn.py [hunk 2]
```diff
@@ -2298,7 +2298,7 @@ def aten_upsample_bilinear2d(
     output_size: Optional[INT64] = None,
     scales_h: Optional[float] = None,
     scales_w: Optional[float] = None,
-    align_corners: bool = True,
+    align_corners: bool = False,
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/nn.py:2301]: Looks like align_corners should be required in the signature

Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2301]: I don't know why it was written in this way. Maybe something to do with onnxscript signature syntax?

File: onnxscript/function_libs/torch_lib/ops/nn.py (hunk 1 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2312]: Unclear to me, looks like different scale values are not covered in tests.

File: onnxscript/function_libs/torch_lib/ops/nn.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2301]: I don't think it is constrained by onnxscript. Looks like something to be fixed

File: onnxscript/function_libs/torch_lib/ops/nn.py (hunk 1 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2312]: @xiaowuhu

File: onnxscript/function_libs/torch_lib/ops/nn.py (hunk 2 above)
Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2301]: Hmm, looks like tests start to fail with the default value swap.

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2301]: The issue is also we are using the tests for `torch.nn.Upsample`. We should probably take the test cases out to create tests for the aten op like @xiaowuhu suggested.

Comment (BowenBao) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2301]: Hmm ok, so it appears the reference test case is for [torch.nn.functional.upsample_bilinear](https://pytorch.org/docs/stable/generated/torch.nn.functional.upsample_bilinear.html#torch.nn.functional.upsample_bilinear), which is deprecated and defaults `align_corners` to True.

I agree the best fix is to make `align_corners` required, and fix the input wrangler.

**Comment by BowenBao** [comment]
> Do you have suggestions? Currently we rely on tests in PyTorch for integration tests.

Based on current infra, one would need to add both test cases to onnxscript and pytorch ... Otherwise, the wrangler itself and operator signature implementation part is missing coverage.

File: onnxscript/function_libs/torch_lib/ops/nn.py [hunk 3]
```diff
@@ -2295,38 +2295,45 @@ def aten_upsample_bicubic2d_backward(
...
 def aten_upsample_bilinear2d_vec(
     self: TReal,
-    output_size: Optional[INT64] = None,
-    align_corners: bool = True,
-    scale_factors: Optional[Sequence[float]] = None,
+    output_size: Optional[INT64],
+    align_corners: bool,
+    scale_factors: Optional[Sequence[float]],
 ) -> TReal:
+    """upsample_bilinear2d(Tensor self, SymInt[2] output_size, bool align_corners, float? scales_h=None, float? scales_w=None) -> Tensor"""
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/nn.py]:
```suggestion
    """upsample_bilinear2d.vec(Tensor input, SymInt[]? output_size, bool align_corners, float[]? scale_factors) -> Tensor"""
```

File: onnxscript/function_libs/torch_lib/ops/nn.py [hunk 4]
```diff
@@ -2295,38 +2295,45 @@ def aten_upsample_bicubic2d_backward(
 @torch_op("aten::upsample_bilinear2d", trace_only=True)
 def aten_upsample_bilinear2d(
     self: TReal,
-    output_size: Optional[INT64] = None,
+    output_size: Optional[INT64],
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/nn.py:2298]:
```suggestion
    output_size: INT64,
```

File: onnxscript/function_libs/torch_lib/ops/nn.py [hunk 5]
```diff
@@ -2295,38 +2295,45 @@ def aten_upsample_bicubic2d_backward(
...
 
+    coordinate_transformation_mode = "align_corners" if align_corners else "pytorch_half_pixel"
     if output_size is not None:
-        result = _aten_upsample_bilinear2d_output_size(self, output_size)
+        result = _aten_upsample_bilinear2d_output_size(
+            self, output_size, coordinate_transformation_mode
+        )
     else:
         assert scales_h is not None
         assert scales_h == scales_w, f"scale_h({scales_h}) != scale_w({scales_w})"
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/nn.py:2312]: Should we reverse the checks - omit output_size when the scales are not None?

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2312]: The reference here shows scale is considered first if present: https://github.com/pytorch/pytorch/blob/f6767244cf292cb5e434303fe62e576b456e92eb/torch/_decomp/decompositions.py#L3311-L3315

Comment (justinchuby) [reply onnxscript/function_libs/torch_lib/ops/nn.py:2312]: But testing suggested otherwise

**Pull request merged**